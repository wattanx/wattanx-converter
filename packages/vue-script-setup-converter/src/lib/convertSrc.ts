import {
  CallExpression,
  ScriptTarget,
  SyntaxKind,
  Project,
  Node,
  ts,
} from "ts-morph";
import { parse } from "@vue/compiler-sfc";
import { getNodeByKind } from "./helper";
import { convertPageMeta } from "./converter/pageMetaConverter";
import { convertProps } from "./converter/propsConverter";
import { convertSetup } from "./converter/setupConverter";
import { convertEmits } from "./converter/emitsConverter";

export const convertSrc = (input: string) => {
  const {
    descriptor: { script },
  } = parse(input);

  const project = new Project({
    compilerOptions: {
      target: ScriptTarget.Latest,
    },
    useInMemoryFileSystem: true,
  });

  const sourceFile = project.createSourceFile("s.tsx", script?.content ?? "");
  const lang = script?.lang ?? "js";

  const callexpression = getNodeByKind(sourceFile, SyntaxKind.CallExpression);

  if (!callexpression) {
    throw new Error("defineComponent is not found.");
  }
  if (!Node.isCallExpression(callexpression)) {
    throw new Error("defineComponent is not found.");
  }

  if (!isDefineComponent(callexpression)) {
    throw new Error("defineComponent is not found.");
  }

  const pageMeta = convertPageMeta(callexpression, lang) ?? "";
  const props = convertProps(callexpression, lang) ?? "";
  const emits = convertEmits(callexpression, lang) ?? "";
  const statement = convertSetup(callexpression) ?? "";

  const statements = project.createSourceFile("new.tsx");

  statements.addStatements(
    sourceFile
      .getStatements()
      .filter((state) => !Node.isExportAssignment(state))
      .map((x) => x.getText())
  );

  if (isDefineNuxtComponent(callexpression)) {
    statements.addStatements(pageMeta);
  }

  statements.addStatements(props);
  statements.addStatements(emits);
  statements.addStatements(statement);

  statements.formatText({
    semicolons: ts.SemicolonPreference.Insert,
    indentSize: 2,
  });

  return statements.getFullText();
};

const isDefineComponent = (node: CallExpression) => {
  if (!Node.isIdentifier(node.getExpression())) {
    return false;
  }

  return (
    node.getExpression().getText() === "defineComponent" ||
    node.getExpression().getText() === "defineNuxtComponent"
  );
};

const isDefineNuxtComponent = (node: CallExpression) => {
  if (!Node.isIdentifier(node.getExpression())) {
    return false;
  }
  return node.getExpression().getText() === "defineNuxtComponent";
};
