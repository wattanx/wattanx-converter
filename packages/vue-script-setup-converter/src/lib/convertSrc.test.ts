import { expect, describe, it } from "vitest";
import { convertSrc } from "./convertSrc";

describe("snapshot", () => {
  it("lang=js", () => {
    const output = convertSrc(`<script>
import { defineComponent, toRefs, computed, ref } from 'vue';

export default defineComponent({
  name: 'HelloWorld',
  emits: {
    change: null
  },
  props: {
    msg: {
      type: String,
      default: 'HelloWorld'
    },
    foo: {
      type: String,
      required: true,
    }
  },
  setup(props, ctx) {
    const { msg, foo } = toRefs(props);
    const newMsg = computed(() => msg.value + '- HelloWorld');

    const count = ref(0);

    return {
      count,
      newMsg
    }
  }
})
</script>`);
    expect(output).toMatchSnapshot();
  });

  it("lang=ts", () => {
    const output = convertSrc(`<script lang="ts">
import { defineComponent, toRefs, computed, ref } from 'vue';

export default defineComponent({
  name: 'HelloWorld',
  emits: {
    change: (value: number) => true,
  },
  props: {
    msg: {
      type: String,
      default: 'HelloWorld'
    },
    foo: {
      type: String
      required: true,
    }
  },
  setup(props, ctx) {
    const { msg, foo } = toRefs(props);
    const newMsg = computed(() => msg.value + '- HelloWorld');

    const count = ref(0);
    ctx.emit("change", 124);

    return {
      count,
      newMsg
    }
  }
})
</script>`);
    expect(output).toMatchSnapshot();
  });

  it("defineNuxtComponent", () => {
    const output = convertSrc(`<script lang="ts">
import { defineNuxtComponent, useNuxtApp } from '#imports';

export default defineNuxtComponent({
  name: 'HelloWorld',
  layout: 'test-layout',
  middleware: 'test-middleware',
  setup(props, ctx) {
    const { $client } = useNuxtApp();

    const onSubmit = () => {
      console.log('onSubmit')
    }

    return {
      onSubmit,
    }
  }
})
</script>
`);
    expect(output).toMatchSnapshot();
  });

  it("empty setup context", () => {
    const output = convertSrc(`<script lang="ts">
import { defineComponent, toRefs, computed } from 'vue';

export default defineComponent({
  name: 'HelloWorld',
  props: {
    msg: {
      type: String,
      default: 'HelloWorld'
    }
  },
  setup(props) {
    const { msg } = toRefs(props);
    const newMsg = computed(() => msg.value + '- HelloWorld');

    return {
      newMsg
    }
  }
})
</script>`);
    expect(output).toMatchInlineSnapshot(
      `
      "import { defineComponent, toRefs, computed } from 'vue';
      type Props = { msg?: string; }; const props = withDefaults(defineProps<Props>(), { msg: 'HelloWorld' });

      const { msg } = toRefs(props);
      const newMsg = computed(() => msg.value + '- HelloWorld');
      "
    `
    );
  });

  it("props no toRefs", () => {
    const output = convertSrc(`<script lang="ts">
    import type { PropType } from 'vue';
import { defineComponent, computed } from 'vue';

export default defineComponent({
  name: 'HelloWorld',
  props: {
    msg: {
      type: String,
      default: 'HelloWorld'
    }
  },
  setup(props) {
    const newMsg = computed(() => props.msg + '- HelloWorld');

    return {
      newMsg
    }
  }
})
</script>`);
    expect(output).toMatchInlineSnapshot(
      `
      "import type { PropType } from 'vue';
      import { defineComponent, computed } from 'vue';
      type Props = { msg?: string; }; const props = withDefaults(defineProps<Props>(), { msg: 'HelloWorld' });

      const newMsg = computed(() => props.msg + '- HelloWorld');
      "
    `
    );
  });
});
