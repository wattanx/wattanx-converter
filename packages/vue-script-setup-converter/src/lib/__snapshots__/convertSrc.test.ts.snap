// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`snapshot > defineNuxtComponent 1`] = `
"import { defineNuxtComponent, useNuxtApp } from '#imports';
definePageMeta({
  name: 'HelloWorld', layout: 'test-layout', middleware: 'test-middleware'
});

const { $client } = useNuxtApp();

const onSubmit = () => {
  console.log('onSubmit');
};
"
`;

exports[`snapshot > lang=js 1`] = `
"import { defineComponent, toRefs, computed, ref } from 'vue';
const props = defineProps({
  msg: {
    type: String,
    default: 'HelloWorld'
  }, foo: {
    type: String,
    required: true,
  }
});
const emit = defineEmits({ change: null });

const { msg, foo } = toRefs(props);
const newMsg = computed(() => msg.value + '- HelloWorld');

const count = ref(0);
"
`;

exports[`snapshot > lang=ts 1`] = `
"import { defineComponent, toRefs, computed, ref } from 'vue';
type Props = {
  msg?: string;
  foo: string;
}; const props = withDefaults(defineProps<Props>(), { msg: 'HelloWorld' });
const emit = defineEmits<{ (e: 'change', value: number): void; }>();

const { msg, foo } = toRefs(props);
const newMsg = computed(() => msg.value + '- HelloWorld');

const count = ref(0);
emit("change", 124);
"
`;
